BJS2-27077   [Experienced] Проверка постов на оскорбительный контент

kripersh27 opened the pull request:
Добавлена функциональность проверки постов на оскорбительный контент

chudoudoribakit committed:
Добавил QuartzConfig, TestAsyncConfig
Добавил классы в сервисный слой для функциональности проверки на оскорбительный контент постов.
Добавил файл forbidden-words.txt со списком оскорбительных слов.

kripersh27 committed:
Merge branch 'chimera-master-stream6' into feature-BJS2-27077

IvanTsegelnyuk commented on src/main/java/faang/school/postservice/service/post/ModerationPostService.java:
```diff
@@ -0,0 +1,58 @@
+    @Value("${moderation.sublist-size}")
+    private int sublistSize;
+
+    @Async
+    @Transactional
```
Async и Transactional же не сработает, т.к. метод вызывается внутри другого метода этого же класса. Проксирование не сработает. Нужно, по идее, выносить в отдельный класс этот метод

sultan-inked replied:
GPT согласен с Иваном:
*Да, комментарий верный. Проблема в том, что аннотации @Async и @Transactional полагаются на механизм проксирования Spring, который работает только при вызове методов через прокси-объект (то есть извне самого класса).*
*Когда вы вызываете метод, аннотированный @Async или @Transactional, из другого метода того же класса, Spring не применяет прокси для этого вызова, потому что это внутренний вызов. Поэтому ни асинхронность, ни транзакционность не будут работать.*

kripersh27 replied:
Да, я уже переделываю, спасибо, ребят)
Лекцию у Влада не досмотрел и не узнал об этом

kripersh27 committed:
Merge branch 'chimera-master-stream6' into feature-BJS2-27077

chudoudoribakit committed:
BJS2-27077 [Experienced] Проверка постов на оскорбительный контент

Правки после кросс ревью, разделил транзакшнл и асинк, поправил ченджсеты, исправил тесты

chudoudoribakit committed:
Merge remote-tracking branch 'origin/feature-BJS2-27077' into feature-BJS2-27077

kripersh27 committed:
Merge branch 'chimera-master-stream6' into feature-BJS2-27077

chudoudoribakit committed:
BJS2-27077 [Experienced] Проверка постов на оскорбительный контент

Вынес магические числа из конфигурации в апликейшн ямл

chudoudoribakit committed:
Merge remote-tracking branch 'origin/feature-BJS2-27077' into feature-BJS2-27077

chudoudoribakit committed:
BJS2-27077 [Experienced] Проверка постов на оскорбительный контент

Небольшие изменения

chudoudoribakit committed:
BJS2-27077  [Experienced] Проверка постов на оскорбительный контент

Небольшие исправления

chudoudoribakit committed:
BJS2-27077  [Experienced] Проверка постов на оскорбительный контент

Убрал использование паралельного стрима

chudoudoribakit committed:
BJS2-27077  [Experienced] Проверка постов на оскорбительный контент

Небольшие изменения

20cognosce commented on src/main/java/faang/school/postservice/config/QuartzConfig.java:
```diff
@@ -0,0 +1,40 @@
+
+@Configuration
+@EnableScheduling
+public class QuartzConfig {
```
Можно спокойно без него обойтись

kripersh27 replied:
Хотелось попробовать

20cognosce commented on src/main/java/faang/school/postservice/config/QuartzConfig.java:
```diff
@@ -0,0 +1,40 @@
+                .repeatForever();
+
+        return TriggerBuilder.newTrigger()
+                .forJob(moderationJobDetail())
```
Ты создаешь два инстанса
Один - бин, второй - вызовом метода

20cognosce commented on src/main/java/faang/school/postservice/model/Post.java:
```diff
@@ -82,4 +88,11 @@ public class Post {
     @Column(name = "updated_at")
     private LocalDateTime updatedAt;
+
+    @Enumerated(EnumType.STRING)
+    @Column(name = "verification_status", nullable = false)
+    private VerificationPostStatus verificationStatus = VerificationPostStatus.UNVERIFIED;
```
Не понял, почему ты на уровне приложения чет присваиваешь

20cognosce commented on src/main/java/faang/school/postservice/repository/PostRepository.java:
```diff
@@ -26,4 +26,7 @@ public interface PostRepository extends JpaRepository<Post, Long> {
     @Query("SELECT p FROM Post p WHERE p.published = false AND p.deleted = false AND p.scheduledAt <= CURRENT_TIMESTAMP")
     List<Post> findReadyToPublish();
+
+    @Query("SELECT p FROM Post p WHERE p.verificationStatus = 'UNVERIFIED'")
+    List<Post> findUnverifiedPosts();
```
А можно не писать запросы на каждый метод, а юзать спринговый нейминг методов, по которым Spring Data сама составит query

20cognosce commented on src/main/java/faang/school/postservice/service/post/ModerationDictionary.java:
```diff
@@ -0,0 +1,49 @@
+
+    @PostConstruct
+    public void loadForbiddenWords() {
+        try (InputStream inputStream = getClass().getResourceAsStream("/forbidden-words.txt");
```
не txt
json, csv

20cognosce replied:
Используй спринговый класс для чтения из ресурсов, не пиши столько лишнего кода

20cognosce replied:
ClassPathResource или ResourceUtil.getFile

20cognosce replied:
Почему не в конструкторе?

kripersh27 replied:
Пост конструкт гарантирует, что бин и зависимости будут полностью инициализированы перед тем, как произойдет загрузка слов из файла, и если вдруг ошибка будет какая-то, то конструктор сломается

20cognosce commented on src/main/java/faang/school/postservice/service/post/ModerationJob.java:
```diff
@@ -0,0 +1,22 @@
+@Component
+public class ModerationJob implements Job {
+
+    @Autowired
```
Autowired нежелательно

20cognosce commented on src/main/java/faang/school/postservice/service/post/ModerationPostService.java:
```diff
@@ -0,0 +1,45 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class ModerationPostService {
```
Сколько у тебя сервисов, почему так много?
Достаточно 3 - джоба, словарь, модератор

Джоба берет все неверифицированные посты, отправляет по подсписку асинхронно в модератор, модератор с помощью словаря в транзакции проверяет каждый пост из подписка, всё

Чем лишние бины городить - убирай тогда @Async и через тред пул в джобе сделай

20cognosce commented on src/main/java/faang/school/postservice/service/post/ModerationPostService.java:
```diff
@@ -0,0 +1,45 @@
+        List<List<Post>> sublists = splitListIntoSublists(unverifiedPosts, sublistSize);
+        log.info("Разделение на подгруппы для модерации, размер каждой группы: {}", sublistSize);
+
+        sublists.forEach(moderationAsyncService::moderatePostsSublistAsync);
```
Ты внутри транзакции вызываешь асинхронный метод, зачем тебе транзакция, если в ней ничего не обновляется, она фактически рид онли

20cognosce replied:
moderatePostsSublistAsync выполняется в другом потоке и вообще без транзакции

20cognosce replied:
Переделать всё

20cognosce commented on src/main/resources/db/changelog/changeset/post_V005_alter_post.sql:
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE post
+    ADD COLUMN verification_status VARCHAR(50) NOT NULL DEFAULT 'UNVERIFIED';
```
Вот ты же итак дефолтное назначаешь, нафига в энтити присвоение

20cognosce commented on src/main/resources/forbidden-words.txt:
```diff
@@ -0,0 +1,2 @@
+badword
+anotherbadword
```
Сделай больше слов, и выбери реальные
Но более менее приличные

fool
idiot
stupid
retard

Вот так достаточно

Русские слова тоже нужны, на двух языках проверяй

chudoudoribakit committed:
BJS2-27077  [Experienced] Проверка постов на оскорбительный контент

Изменения после ревью(убрал асинк, сделал через пул потоков)

20cognosce commented on src/main/java/faang/school/postservice/service/post/ModerationJob.java:
```diff
@@ -0,0 +1,62 @@
+        List<List<Post>> sublists = moderationService.splitListIntoSublists(unverifiedPosts, sublistSize);
+        log.info("Split posts into {} sublists", sublists.size());
+
+        ExecutorService executorService = Executors.newFixedThreadPool(threadPoolSize);
```
Почему ты на каждое срабатывание джобы новый экзекьютор создаешь...

20cognosce commented on src/main/java/faang/school/postservice/service/post/ModerationJob.java:
```diff
@@ -0,0 +1,62 @@
+            for (List<Post> sublist : sublists) {
+                executorService.submit(() -> moderationService.moderatePostsSublist(sublist));
+            }
+        } finally {
+            executorService.shutdown();
+            try {
+                if (!executorService.awaitTermination(awaitTerminationSeconds, TimeUnit.SECONDS)) {
+                    log.warn("Executor service did not terminate in the specified time.");
+                    List<Runnable> droppedTasks = executorService.shutdownNow();
+                    log.warn("Dropped {} tasks", droppedTasks.size());
+                }
+            } catch (InterruptedException e) {
+                executorService.shutdownNow();
+                Thread.currentThread().interrupt();
+            }
```
Да зачем, пусть работает в течение всего рантайма

20cognosce replied:
Ты ведь если сабмитишь подсписок, и все потоки заняты, то там будет очередь тасок и они будут просто ждать свободного потока

20cognosce commented on src/main/resources/forbidden-words.json:
```diff
@@ -0,0 +1,12 @@
+[
```
😂

20cognosce reviewed (changes requested):
Пул потоков поправь и можешь мерджить

chudoudoribakit committed:
BJS2-27077  [Experienced] Проверка постов на оскорбительный контент

Изменения после ревью(вынес экзекьютор сервис в поля, чтобы создавался один раз, и убрал его закрытие - работать будет до конца программы)

chudoudoribakit committed:
BJS2-27077  [Experienced] Проверка постов на оскорбительный контент

Небольшие изменения

kripersh27 committed:
Merge branch 'chimera-master-stream6' into feature-BJS2-27077
